File: templates/graph/topological_sort.go
package graph

func TopoSortKahn(numNodes int, edges [][2]int) ([]int, bool) {
	graph := make([][]int, numNodes)
	indegree := make([]int, numNodes)

	for _, e := range edges {
		from, to := e[0], e[1]
		graph[from] = append(graph[from], to)
		indegree[to]++
	}

	queue := []int{}
	for i := 0; i < numNodes; i++ {
		if indegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	order := make([]int, 0, numNodes)
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		order = append(order, node)

		for _, nei := range graph[node] {
			indegree[nei]--
			if indegree[nei] == 0 {
				queue = append(queue, nei)
			}
		}
	}

	return order, len(order) == numNodes
}

File: templates/graph/topological_sort.py
from collections import deque


def topo_sort_kahn(num_nodes: int, edges: list[tuple[int, int]]) -> tuple[list[int], bool]:
    graph = [[] for _ in range(num_nodes)]
    indegree = [0] * num_nodes

    for src, dst in edges:
        graph[src].append(dst)
        indegree[dst] += 1

    queue = deque(i for i in range(num_nodes) if indegree[i] == 0)
    order: list[int] = []

    while queue:
        node = queue.popleft()
        order.append(node)
        for nei in graph[node]:
            indegree[nei] -= 1
            if indegree[nei] == 0:
                queue.append(nei)

    return order, len(order) == num_nodes
